tuf/api/metadata.py

```diff
@@ -0,0 +1,250 @@
+# Imports.
+
+# 1st-party.
+from tuf.api.keys import KeyRing
+
+# 2nd-party.
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta
+from typing import Any, Dict, List, Optional
+
+import json
+import tempfile
+
+# 3rd-party.
+
+from securesystemslib.formats import encode_canonical
+from securesystemslib.util import load_json_file, persist_temp_file
+from securesystemslib.storage import StorageBackendInterface
+from tuf.repository_lib import (
+    _get_written_metadata,
+    _strip_version_number,
+    generate_snapshot_metadata,
+    generate_targets_metadata,
+    generate_timestamp_metadata,
+)
+
+import iso8601
+import tuf.formats
+
+# Types.
+
+JsonDict = Dict[str, Any]
+
+# Classes.
+
+class Metadata(ABC):
+    # By default, a Metadata would be a rather empty one.
+    def __init__(self, consistent_snapshot: bool = True, expiration: datetime = datetime.today(), keyring: Optional[KeyRing] = None, version: int = 1) -> None:
+        self.__consistent_snapshot = consistent_snapshot
+
+        self.__keyring = keyring
+        self.__expiration = expiration
+
+        assert version >= 1, f'{version} < 1'
+        self.__version = version
+
+        self.__signed = {}
+        self.__signatures = []
+
+    # And you would use this method to populate it from a file.
+    @classmethod
+    def read_from_json(cls, filename: str, storage_backend: Optional[StorageBackendInterface] = None) -> Metadata:
+        signable = load_json_file(filename, storage_backend)
+        tuf.formats.SIGNABLE_SCHEMA.check_match(signable)
+
+        signatures = signable['signatures']
+        signed = signable['signed']
+
+        # We always intend times to be UTC
+        # NOTE: we could do this with datetime.fromisoformat() but that is not
+        # available in Python 2.7's datetime
+        expiration = iso8601.parse_date(signed['expires']).replace(tzinfo=None)
+        version = signed['version']
+
+        fn, fn_ver = _strip_version_number(filename, True)
+        if fn_ver:
+            assert fn_ver == self.__version, f'{fn_ver} != {self.__version}'
+            consistent_snapshot = True
+        else:
+            consistent_snapshot = False
+
+        metadata = cls(consistent_snapshot=consistent_snapshot,
+                       expiration=expiration,
+                       version=version)
+
+        metadata._signatures = signatures
+        metadata._signed = signed
+
+        return metadata
+
+    @property
+    def signable(self) -> JsonDict:
+        return {"signatures": self.signatures,
+                "signed": self.signed}
+
+    @property
+    def signed_bytes(self) -> bytes:
+        return encode_canonical(self.signed).encode('UTF-8')
+
+    @property
+    @abstractmethod
+    def signed(self) -> JsonDict:
+        raise NotImplementedError
```

Comment:
Do we really need abc/abstractmethod/NotImplementedError, etc... in our Metadata base class? IIUC the list of possible subclasses is: root, snapshot, timestamp and targets. I don't think we need to plan for users implementing arbitrary subclasses.

Reply:
It was just a way to denote that this method must be overridden